compiler/cpp/test/ReaderTest.cpp

```diff
@@ -0,0 +1,234 @@
+
+// -------------------------------------------------------- EOF --------------------------------------------------------
+
+TEST(ReaderTests, EoF) {
+    // Initialize
+    Reader reader = Reader("Test", 2);
+
+    // Test
+    EXPECT_EQ (reader.getLookahead(),  'T');
+    reader.advance();
+    reader.advance();
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  't');
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+}
+
+TEST(ReaderTests, EoFInEmptyFile) {
+    // Initialize
+    Reader reader = Reader("", 5);
+
+    // Test
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+}
+
+// ----------------------------------------------------- Lookahead -----------------------------------------------------
+
+TEST(ReaderTests, Lookahead) {
+    // Initialize
+    Reader reader = Reader("File input", 2);
+
+    // Test
+    EXPECT_EQ (reader.getLookahead(),  'F');
+}
+
+TEST(ReaderTests, LookaheadAdvance) {
+    // Initialize
+    Reader reader = Reader("File input", 2);
+
+    // Test
+    EXPECT_EQ (reader.getLookahead(),  'F');
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  'i');
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  'l');
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  'e');
+}
+
+// --------------------------------------------------- Max lookahead ---------------------------------------------------
+
+TEST(ReaderTests, MaxLookahead) {
+    // Initialize
+    Reader reader = Reader("File input", 4);
+
+    // Test
+    EXPECT_EQ (reader.getMaxLookahead(),  "File");
+}
+
+TEST(ReaderTests, MexLookaheadAdvance) {
+    // Initialize
+    Reader reader = Reader("File input", 3);
+
+    // Test
+    EXPECT_EQ (reader.getMaxLookahead(),  "Fil");
+    reader.advance();
+    EXPECT_EQ (reader.getMaxLookahead(),  "ile");
+    reader.advance();
+    EXPECT_EQ (reader.getMaxLookahead(),  "le ");
+    reader.advance();
+    EXPECT_EQ (reader.getMaxLookahead(),  "e i");
+}
+
+TEST(ReaderTests, MaxLookaheadInvalid) {
+    // Initialize & test
+    try {
+        Reader reader = Reader("Test", 0);
+        FAIL() << "Expected MaxLookaheadException";
+    } catch(MaxLookaheadException const & err) {
+        EXPECT_EQ(err.what(),std::string("Invalid maximum lookahead '0' specified. Must be >= 1"));
+    }
+}
+
+TEST(ReaderTests, MaxLookaheadIsOne) {
+    // Initialize
+    Reader reader = Reader("File input", 1);
+
+    // Test
+    EXPECT_EQ (reader.getMaxLookahead(),  "F");
+    EXPECT_EQ (reader.getMaxLookahead(),  std::string(1, reader.getLookahead()));
+    reader.advance();
+    reader.advance();
+    reader.advance();
+    EXPECT_EQ (reader.getMaxLookahead(),  "e");
+    EXPECT_EQ (reader.getMaxLookahead(),  std::string(1, reader.getLookahead()));
+}
+
```

Comment:
I'm missing

- a test for look ahead and expect with line breaks
- a test for "Look ahead and expect with two characters" where the max look ahead is bigger than the file, e.g. take the file `{a` and a max look ahead of `4`. Then also check that `lookAheadMultiple()` includes EOF like here:

https://github.com/compose-generator/ccom/blob/dee50ea77063855207c5475e2f79029947cf1e04/compiler/java/src/test/java/com/c/reader/ReaderTest.java#L234-L251

Reply:
Do you return EOF as "fill up char"? My reader in this case only returns "{a" without any EOFs in the end. This is a better approach in my opinion. What do you think?

Reply:
Added the two requested tests. Should be fine now ...

Reply:
@marcauberer Yes, I do use EOF as "fill up char". I terat the input as a stream of chars and after that infinite EOFs. So after heaving read the whole file, you can still `advance()` and read another EOF...

I guess your reader implementation is fine as well. However, it might make it more difficult for the users of the Reader to use the `lookAheadMultiple()` method, e.g. a user of the reader always has to check if the `lookAheadMultiple()` method really returned 4 chars (when max look ahead is 4). That's why I guarantee in my Reader that the number of chars returned is **always** equal to the max look ahead passed to the Reader upon construction.

Reply:
Seems fair enough, but I leave it like it is for now okay?

Reply:
@marcauberer ok, maybe you can put this on your Kanban board to evaluate this again in the future?